logo.js

```diff
@@ -924,84 +924,38 @@ function LogoInterpreter(turtle, stream, savehook)
     self.forceBye = true;
   };
 
-  self.run = function(string, options) {
-    if (self._pendingPromise) {
-      var runner = self.run.bind(self, string, options);
-      return self._queueCall(runner);
-    }
-    options = Object(options);
-    // Parse it
-    var atoms = parse(string);
-
-    if (self.turtle) { self.turtle.begin(); }
-    // And execute it!
-    var promise = self.execute(atoms, options).then(function (result) {
-      if (self.turtle) {
-        self.turtle.end();
-      }
-      return result;
-    }, function (err) {
-      if (self.turtle) {
-        self.turtle.end();
-      }
-      if (err && (err.special == "bye" || err.special == "stop")) {
-        return undefined;
-      }
-      throw err;
-    });
-    this._setPendingPromise(promise);
-    self._pendingPromise = promise;
-    return promise;
-  };
+  function noop() {}
+  var lastRun = Promise.resolve();
 
-  //----------------------------------------------------------------------
-  // Helpers to ensure calls to run() are serialized:
-  //----------------------------------------------------------------------
+  self.run = function(string, options) {
+    var promise = lastRun.then(function() {
```

Comment:
I'm pretty sure this was my first approach, except it didn't mean that calls to self.run would be run in order, or it created a really long promise chain that I noticed cause performance problems.  But this was all with the tests, are you seeing the tests run okay even when .run() gets called a bunch of times in a row?